cli/text_encoding.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+use encoding_rs::*;
+use std::{
+  borrow::Cow,
+  io::{Error, ErrorKind},
+};
+
+/// Attempts to detect the character encoding of the provided bytes.
+///
+/// Supports UTF-8, UTF-16 Little Endian and UTF-16 Big Endian.
+pub fn detect_charset(bytes: &[u8]) -> &str {
+  const UTF8_BOM: &[u8] = b"\xEF\xBB\xBF";
+  const UTF16_LE_BOM: &[u8] = b"\xFF\xFE";
+  const UTF16_BE_BOM: &[u8] = b"\xFE\xFF";
+
+  if bytes.starts_with(UTF8_BOM) {
+    "utf-8"
+  } else if bytes.starts_with(UTF16_LE_BOM) {
+    "utf-16le"
+  } else if bytes.starts_with(UTF16_BE_BOM) {
+    "utf-16be"
+  } else {
+    // Assume everything else is utf-8
+    "utf-8"
+  }
+}
+
+/// Attempts to convert the provided bytes to a UTF-8 string.
+///
+/// Supports all encodings supported by the encoding crate, which includes all encodings specified in the WHATWG Encoding Standard (see: https://encoding.spec.whatwg.org/).
```

Comment:
I would suggest to support *only* the charsets specified by the WHATWG. Because all charsets that we support now become public API and have to be kept forever.

Reply:
@piscisaureus I've checked and my comment is incorrect: the `encoding_rs` crate we're using supports _only_ the charsets specified by the WHATWG. They even mention this in their [CONTRIBUTING.md](https://github.com/hsivonen/encoding_rs/blob/master/CONTRIBUTING.md#no-encodings-beyond-the-encoding-standard)

Question now is: can we rely on their stance being held over time, or do we want to be cautious and filter the charsets according to the WHATWG spec by ourselves?

Reply:
Fixed

Reply:
> do we want to be cautious and filter the charsets according to the WHATWG spec by ourselves?

If the crate already explicitly says that it only supports WHATWG encodings there's no need to filter ourselves IMO.

Reply:
Alright, updated the code comment.